airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/state/StateManager.kt

```diff
@@ -44,50 +39,73 @@ interface StateManager {
     )
 
     private val stateIsGlobal: AtomicReference<Boolean?> = AtomicReference(null)
     private val streamStates:
-        ConcurrentHashMap<DestinationStream, LinkedHashMap<Long, DestinationStateMessage>> =
+        ConcurrentHashMap<DestinationStream, ConcurrentLinkedHashMap<Long, T>> =
         ConcurrentHashMap()
-    private val globalStates: ConcurrentLinkedQueue<GlobalState> = ConcurrentLinkedQueue()
-
-    override fun addStreamState(
-        stream: DestinationStream,
-        index: Long,
-        stateMessage: DestinationStateMessage
-    ) {
-        if (stateIsGlobal.getAndSet(false) != false) {
+    private val globalStates: ConcurrentLinkedQueue<GlobalState<T>> = ConcurrentLinkedQueue()
+
+    override fun addStreamState(key: DestinationStream, index: Long, stateMessage: T) {
+        if (stateIsGlobal.updateAndGet { it == true } != false) {
             throw IllegalStateException("Global state cannot be mixed with non-global state")
         }
 
-        val streamStates = streamStates.getOrPut(stream) { LinkedHashMap() }
-        streamStates[index] = stateMessage
-        log.info { "Added state for stream: $stream at index: $index" }
+        streamStates.compute(key) { _, indexToMessage ->
+            val map =
+                if (indexToMessage == null) {
+                    // If the map doesn't exist yet, build it.
+                    ConcurrentLinkedHashMap.Builder<Long, T>().maximumWeightedCapacity(1000).build()
+                } else {
+                    if (indexToMessage.isNotEmpty()) {
+                        // Make sure the messages are coming in order
+                        val oldestIndex = indexToMessage.ascendingKeySet().first()
+                        if (oldestIndex > index) {
+                            throw IllegalStateException(
+                                "State message received out of order ($oldestIndex before $index)"
+                            )
+                        }
+                    }
+                    indexToMessage
+                }
+            // Actually add the message
+            map[index] = stateMessage
+            map
+        }
+
+        log.info { "Added state for stream: $key at index: $index" }
     }
 
-    override fun addGlobalState(
-        streamIndexes: List<Pair<DestinationStream, Long>>,
-        stateMessage: DestinationStateMessage
-    ) {
-        if (stateIsGlobal.getAndSet(true) != true) {
+    // TODO: Is it an error if we don't get all the streams every time?
+    override fun addGlobalState(keyIndexes: List<Pair<DestinationStream, Long>>, stateMessage: T) {
+        if (stateIsGlobal.updateAndGet { it != false } != true) {
             throw IllegalStateException("Global state cannot be mixed with non-global state")
         }
 
-        globalStates.add(GlobalState(streamIndexes, stateMessage))
-        log.info { "Added global state with stream indexes: $streamIndexes" }
+        val head = globalStates.peek()
```

Comment:
Actually these could just be channels, I think that's more idiomatic kotlin?

With a compare/swap guard on the max seen to prevent out-of-order (which should never happen anyway)